src/OpenSpaceToolkit/Astrodynamics/Solvers/LeastSquaresSolver.cpp

```diff
@@ -0,0 +1,389 @@
+        {
+            throw ostk::core::error::RuntimeError("Observations must have the same coordinate subsets.");
+        }
+        if (observation.getFrame() != observationBuilder.getFrame())
+        {
+            throw ostk::core::error::RuntimeError("Observations must have the same frame.");
+        }
+    }
+
+    // Get dimensions
+    const Size stateDimension = anInitialGuessState.accessCoordinates().size();
+    const Size observationDimension = anObservationArray[0].accessCoordinates().size();
+    const Size observationCount = anObservationArray.size();
+
+    // Validate dimensions
+    if (observationDimension * observationCount <= stateDimension)
+    {
+        throw ostk::core::error::RuntimeError(
+            "Observation count * observation dimension should be greater than estimated State dimension to "
+            "yield a full rank H matrix."
+        );
+    }
+
+    // Setup a priori and covariance matrices
+    // x̄
+    VectorXd xApriori = VectorXd::Zero(stateDimension);
+
+    // P̄⁻¹ = diag(1/σ²)
+    const MatrixXd PAprioriInverse = anInitialGuessSigmas.empty()
+                                       ? MatrixXd::Zero(stateDimension, stateDimension)
+                                       : extractSigmas(anInitialGuessSigmas, estimationStateBuilder);
+
+    // Setup measurement covariance matrix
+    // R⁻¹ = diag(1/σ²)
+    const MatrixXd RInv = anObservationSigmas.empty() ? MatrixXd::Identity(observationDimension, observationDimension)
+                                                      : extractSigmas(anObservationSigmas, observationBuilder);
+
+    // Initialize state vectors
+    // X∗ (nominal trajectory)
+    VectorXd XNom = anInitialGuessState.accessCoordinates();
+    // x̂ = X - X∗
+    VectorXd xHat = VectorXd::Zero(stateDimension);
+
+    // Get observations coordinates
+    // Y
+    MatrixXd observationCoordinates(observationDimension, observationCount);
+    for (Size i = 0; i < observationCount; ++i)
+    {
+        observationCoordinates.col(i) = anObservationArray[i].inFrame(estimationStateFrame).accessCoordinates();
+    }
+
+    // Initialize arrays for the iteration
+    MatrixXd residuals = MatrixXd::Zero(observationDimension, observationCount);
+    Real previousRmsError = 0.0;
+    String terminationCriteria = "Maximum Iteration Threshold";
+    Array<Step> steps;
+
+    MatrixXd Lambda = MatrixXd::Zero(stateDimension, stateDimension);
+    MatrixXd PHatFrisbee = MatrixXd::Zero(stateDimension, stateDimension);
+
+    const Array<Instant> observationInstants = anObservationArray.map<Instant>(
+        [](const auto& state) -> Instant
+        {
+            return state.getInstant();
+        }
+    );
+
+    State currentEstimatedState = State::Undefined();
+
+    const auto computeObservationsCoordinates = [&](const State& state, const Array<Instant>& instants) -> MatrixXd
+    {
+        const Array<State> states = aStateGenerator(state, instants);
```

Comment:
```suggestion
        const Array<State> computedObservationsStates = aStateGenerator(state, instants);
```

Reply:
I think in the scope of the lambda we don't need to have such verbose naming, the name of the lambda clarifies this and it's a short lambda.